### PR TITLE
[JENKINS-11257]: Aborting the parent job will abort building/queued child jobs

### PR DESCRIPTION
**Pull request by ad22**
Catch InterruptedException, do a cleanup of the queue and building items and then throw a new InterruptedException

### CODE REVIEW

**Comment by ad22** [comment]
The unit tests will fail in this request until https://github.com/jenkinsci/parameterized-trigger-plugin/pull/103 is merged in.

**Comment by kwhetstone** [comment]
Could you add a test to this?

File: src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java [hunk 1]
```diff
@@ -151,6 +160,57 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
...
+                                    }
+                                } catch (javax.servlet.ServletException ex) {
+                                    // pass, build has already completed or aborted
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            throw new InterruptedException();
```
Comment (sschuberth) [inline src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java]: Maybe re-throw with `y.getMessage()` passed to the constructor?

File: src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java [hunk 2]
```diff
@@ -151,6 +160,57 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
                     }
                 }
             }
+        } catch (InterruptedException y) {
```
Comment (sschuberth) [inline src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java]: Maybe rename `y` to the more common `e` here?

**Comment by sschuberth** [comment]
@ad22 Would you be able to add the requested test, as I'm also interested in seeing this merged? If not, I can probably give a hand.

**Comment by ad22** [comment]
@kwhetstone @sschuberth I haven't been able to find time to attend to this. I'll attempt to have a new change up this week. If not, @sschuberth I'll take your help.

**Comment by ad22** [comment]
@kwhetstone @sschuberth I've added tests and a few changes.

**Comment by sschuberth** [comment]
Thanks @ad22! We'll try to test this locally (/cc @gregleno).

In any case, @kwhetstone, can you please review again and probably merge soon?

**Comment by kwhetstone** [comment]
Thanks @ad22; I'll take a look here soon.  I'm looking forward to this feature!

**Comment by sschuberth** [comment]
Hey, @kwhetstone is "soon" yet there? 😉

**Comment by oleg-nenashev** [comment]
I will try to take a look within a couple of weeks

**Comment by sschuberth** [comment]
@ad22 Mind rebasing this one to resolve conflicts so @oleg-nenashev has no excuse not to merge it? 😉

**Comment by dbarbieri** [comment]
This conflict was really easy to resolve locally, just some import statements. No knowledge of the functionality of the patch was necessary.

**Comment by zzhengjian** [comment]
when will this fix be merged, we're kinda waiting for the fix to be released. :)

**Comment by ad22** [comment]
We have been running with this change without too many problems for over a year, but I found a couple of bugs that I fixed in the latest commit:

1. Do not abort jobs with User/Remote cause (The Rebuild plugin, for example, appends the cause to the existing upstream cause)
2. Support abort of pipeline jobs.

I noticed that if #108 was to go in, we will have the exact list of runs as soon as they start, so walking through the builds will probably be redundant. The queue will still have to cleaned up.

**Comment by sschuberth** [comment]
Thanks for your continued work on this, @ad22. I really hope this will get merged finally.

**Comment by sschuberth** [comment]
@oleg-nenashev, any follow-up on [your comment](https://github.com/jenkinsci/parameterized-trigger-plugin/pull/104#issuecomment-310115043).

File: src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java [hunk 3]
```diff
@@ -156,7 +171,62 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
...
+                                    List<Cause> causes = queueItem.getCauses();
+                                    Cause buildQueueCause = null;
+                                    boolean userQueueCause = false;
+                                    for (Cause c : causes) {
+                                        if (c instanceof Cause.UserIdCause || c instanceof Cause.UserCause || c instanceof Cause.RemoteCause) {
+                                            userQueueCause = true;
+                                            break;
+                                        }
+                                        if ((c instanceof Cause.UpstreamCause)) {
+                                            buildQueueCause = c;
```
[the remark this answers is not in the document]
Comment (ad22) [reply src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java]: Thanks @kwhetstone ! I'll make these minor updates in the next few days.

[the remark this answers is not in the document]
Comment (tristanzhang) [reply src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java]: @ad22  Any progress on this issue? I'll be appreciated if this PR can be merged as soon as possible.

**Comment by orgads** [comment]
> Personally I think this looks ok. #108 has conflicts that need to be resolved before merge anyway, so that's not going in for awhile.

Looks like there is no progress for #108. Can this be merged, and then resolve the conflicts in #108?

**Comment by ad22** [comment]
I've rebased to the latest state of the branch and updated the review.

**Comment by joelneubert** [comment]
I found this PR while digging through some tickets in jenkins issues. Is there something that would help get it merged? This meets a use case I'm about to write a script to address...

File: src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java [hunk 4]
```diff
@@ -157,7 +164,59 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
...
-                                throw new AbortException(p.getFullDisplayName() +" aborted.");
+				throw new AbortException(p.getFullDisplayName() +" aborted.");
+			    } catch (InterruptedException y) {
+                                Queue buildQueue = Jenkins.getInstance().getQueue();
+                                for (Queue.Item queueItem : buildQueue.getItems()) {
+                                    List<Cause> causes = queueItem.getCauses();
+                                    Cause buildQueueCause = null;
+                                    boolean userQueueCause = false;
+                                    for (Cause c : causes) {
+                                        if (c instanceof Cause.UserIdCause || c instanceof Cause.UserCause || c instanceof Cause.RemoteCause) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java]: I do mot understand why it is so special to be checked here

File: src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java [hunk 5]
```diff
@@ -157,7 +164,59 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
                                     listener.getLogger().println("Skipping " + ModelHyperlinkNote.encodeTo(p) + ". The project was not triggered by some reason.");
                                 }
                             } catch (CancellationException x) {
-                                throw new AbortException(p.getFullDisplayName() +" aborted.");
+				throw new AbortException(p.getFullDisplayName() +" aborted.");
+			    } catch (InterruptedException y) {
+                                Queue buildQueue = Jenkins.getInstance().getQueue();
+                                for (Queue.Item queueItem : buildQueue.getItems()) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java]: It operates with the queue cache, and hence it won't reliably abort sownstream buds in the case of race conditions

File: src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java [hunk 6]
```diff
@@ -157,7 +164,59 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
...
+                                }
+                                RunList runList = p.getBuilds();
+                                for (Iterator it = runList.iterator(); it.hasNext(); ) {
+                                    Run latestBuild = (Run) it.next();
+                                    List<Cause> buildCauses = latestBuild.getCauses();
+                                    Cause.UpstreamCause buildCause = null;
+                                    boolean userCause = false;
+                                    for (Cause c : buildCauses) {
+                                        if (c instanceof Cause.UserIdCause || c instanceof Cause.UserCause || c instanceof Cause.RemoteCause) {
+                                            userCause = true;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/parameterizedtrigger/TriggerBuilder.java]: Check logic above is different

**Comment by almiroshnich** [comment]
@joelneubert, if you don't already – please, take a look at this gist: https://gist.github.com/almiroshnich/734e31ba46cff18ec7a45b99cc791787

**Comment by sschuberth** [comment]
On a related note, here's an own Jenkins plugin for a similar purpose: https://github.com/thors/hierarchy-killer-plugin

**Comment by ad22** [comment]
This PR was my internal work-around to account for the design gap in the interface (inability to **track** the jobs that it spawned). While it has been working for my use case for ~3+ years (we spawn short running massively parallel jobs), I know its not going to be always reliable or performant for all use cases. I'm essentially running a giant loop on the queue and runs of  a given job.
Note that some of my peers have created a separate message queue to track all the jobs spawned by jenkins so that they can be aborted later. Its rather unfortunate that a not-so-easy-to-fix initial design gap has caused a fair bit of chaos.

**Pull request closed without merging**